keps/sig-auth/267-kubelet-server-certificate-bootstrap-rotation/README.md

```diff
@@ -0,0 +1,292 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubelet uses client certificates to authenticate to kube-apiserver. A kubelet also can use serving certificates. The kubelet itself exposes an https endpoint for certain features. To secure these, the kubelet can do one of:
+
+- use provided key and certificate, via the --tls-private-key-file and --tls-cert-file flags
+- create self-signed key and certificate, if a key and certificate are not provided
+- request serving certificates from the cluster server, via the CSR API
+
+## Motivation
+
+The client certificate provided by TLS bootstrapping is signed, by default, for client auth only, and thus cannot be used as serving certificates, or server auth.
+
+This proposal covers a process for generating a key locally and then issuing a
+Certificate Signing Request to the cluster API server to get an associated
+certificate signed by the cluster Certificate Authority. Also, as certificates
+approach expiration, the same mechanism will be used to request an updated
+certificate.
+
+### Goals
+
+- Allow Nodes/Kubelet to request a certificate to an external authority instead of using self signed certificates during bootstrap.
+- Allow Nodes/Kubelet to request renewals of their certificates to an external authority.
+
+
+## Design Details
+
+Kubernetes v1.8 and higher kubelet implements features for enabling rotation of its client and/or serving certificates.
+
+You can configure the kubelet to rotate its client certificates by creating new CSRs as its existing credentials expire. To enable this feature, user can enable the rotateCertificates field of kubelet configuration file.
+
+The CSR approving controllers implemented in core Kubernetes do not approve node serving certificates for security reasons. To use RotateKubeletServerCertificate operators need to run a custom approving controller, or manually approve the serving certificate requests.
+
+A deployment-specific approval process for kubelet serving certificates should typically only approve CSRs which:
+
+1. are requested by nodes (ensure the spec.username field is of the form system:node:<nodeName> and spec.groups contains system:nodes)
+
+2. request usages for a serving certificate (ensure spec.usages contains server auth, optionally contains digital signature and key encipherment, and contains no other usages)
+
+3. only have IP and DNS subjectAltNames that belong to the requesting node, and have no URI and Email subjectAltNames (parse the x509 Certificate Signing Request in spec.request to verify subjectAltNames)
+
+### Test Plan
+
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+
+##### Unit tests
```

Comment:
hardest part of testing when I looked at GA-ing this was testing of the edge cases. How we retry, how we start kubelet that failed to get certs, etc.

Reply:
this is a feature that has been running in production for more than 20 releases, there is no functionality added in this KEP, this is just adding the missing requirement per sig-auth, a builtin component that can provide the same functionality

Reply:
> How we retry, how we start kubelet that failed to get certs, etc.

what do you mean by start? ` how we start kubelet that failed to get certs`